### PR TITLE
Grist cache shit 

### PR DESCRIPTION
**Pull request by KingBalDoro**
beyond a shit ton of things like balancing everything, scaling gutter, and making the rungGrist array have actual things in it this is ready

### CODE REVIEW

File: src/main/java/com/mraof/minestuck/client/gui/playerStats/EcheladderScreen.java [hunk 1]
```diff
@@ -179,21 +186,62 @@ private void drawLadder(PoseStack poseStack, int currentRung, boolean showLastRu
...
+				fill(poseStack, xOffset + 90, y, xOffset + 236, y + 12, backgrounds.length > rung ? backgrounds[rung] : (textColor ^ 0xFFFFFFFF));
 			} else if(rung == currentRung + 1 && animationCycle == 0)
 			{
 				int bg = ~rand.nextInt(0xFFFFFF);
 				if(backgrounds.length > rung)
 					bg = backgrounds[rung];
 				else if(textColors.length > rung)
 					bg = ~textColors[rung];
-				fill(poseStack, xOffset + 90, y + 10, xOffset + 90 + (int)(146* ClientPlayerData.getRungProgress()), y + 12, bg);
+				fill(poseStack, xOffset + 90, y + 10, xOffset + 90 + (int) (146 * ClientPlayerData.getRungProgress()), y + 12, bg);
```
Comment (SarahKirksey) [inline src/main/java/com/mraof/minestuck/client/gui/playerStats/EcheladderScreen.java:197]: I don't think the first extra space on this line is doing anyone any favors, and it can be removed.
That said, it's not like it's really hurting anything, either.

File: src/main/java/com/mraof/minestuck/client/gui/playerStats/EcheladderScreen.java [hunk 2]
```diff
@@ -179,21 +186,62 @@ private void drawLadder(PoseStack poseStack, int currentRung, boolean showLastRu
...
+					{
+						s = I18n.get("echeladder.rung." + rung);
+					}
+					else
+					{
+						s = "Rung " + (rung + 1);
+					}
+				}
+				mc.font.draw(poseStack, s, xOffset + ladderXOffset -
+						mc.font.width(s) / 2, y + 2, textColor);
```
Comment (SarahKirksey) [inline src/main/java/com/mraof/minestuck/client/gui/playerStats/EcheladderScreen.java:242]: It's probably better to make this one line.

Comment (SarahKirksey) [inline src/main/java/com/mraof/minestuck/client/gui/playerStats/EcheladderScreen.java:242]: It's probably better to make this one line.

File: src/main/java/com/mraof/minestuck/client/gui/playerStats/EcheladderScreen.java [hunk 3]
```diff
@@ -228,8 +276,14 @@ private List<Component> drawEffectIconsAndText(PoseStack poseStack, int currentR
...
 		
+		/**
+		 * here we are rendering the grist cache limit.
+		 * we show the current limit by calling the rungGrist array in grist helper.
+		 * this is seen on string.valueOf(GristHelper...
+		 */
 		mc.font.draw(poseStack, I18n.get(CACHE), xOffset + 24, yOffset + 138, 0x404040);
-		mc.font.draw(poseStack, "Unlimited", xOffset + 26, yOffset + 147, 0x0094FF);
+		mc.font.draw(poseStack, String.valueOf(GristHelper.rungGrist[currentRung]), xOffset + 26,
+				yOffset + 147, 0x0094FF);
```
Comment (SarahKirksey) [inline src/main/java/com/mraof/minestuck/client/gui/playerStats/EcheladderScreen.java:286]: It's worth noting that this is the only change to this file that's directly related to the Grist Gutter, and it may be better to see if you can get the other changes into a different branch/PR.

File: src/main/java/com/mraof/minestuck/client/renderer/entity/GristRenderer.java [hunk 4]
```diff
@@ -26,23 +27,45 @@ public GristRenderer(EntityRendererProvider.Context context)
 	@Override
 	public void render(GristEntity grist, float entityYaw, float partialTicks, PoseStack poseStack, MultiBufferSource bufferIn, int packedLightIn)
 	{
+		double shakeScale = 1.0 / 256.0;
```
Comment (SarahKirksey) [inline src/main/java/com/mraof/minestuck/client/renderer/entity/GristRenderer.java:30]: Maybe this could be moved up to line 19, and made into a static final?

File: src/main/java/com/mraof/minestuck/entity/item/GristEntity.java [hunk 5]
```diff
@@ -59,6 +69,18 @@ public GristEntity(EntityType<? extends GristEntity> type, Level level)
...
 	}
 	
+	/**
+	 * this is where we set up our consumedelay
+	 */
+	public GristEntity(Level level, double x, double y, double z, GristAmount gristData, int pickupDelay)
+	{
+		this(level, x, y, z, gristData);
+		consumeDelay = pickupDelay;
+		// Set the class's consume-delay variable to equal the pickupDelay value that got passed in.
```
Comment (SarahKirksey) [inline src/main/java/com/mraof/minestuck/entity/item/GristEntity.java:79]: Comments describing an action should go above the action being described.

File: src/main/java/com/mraof/minestuck/entity/item/GristEntity.java [hunk 6]
```diff
@@ -199,27 +243,67 @@ protected void readAdditionalSaveData(CompoundTag compound)
 			this.gristType = GristType.read(compound, "Type");
 	}
 	
+	public int shaderAlpha = 0;
```
Comment (SarahKirksey) [inline src/main/java/com/mraof/minestuck/entity/item/GristEntity.java:246]: Member variables should go at the top of the class, like where you have "animation"

File: src/main/java/com/mraof/minestuck/entity/item/GristEntity.java [hunk 7]
```diff
@@ -199,27 +243,67 @@ protected void readAdditionalSaveData(CompoundTag compound)
...
+	
 	/**
 	 * Called by a player entity when they collide with an entity
+	 * dectects collision with player entity and
```
Comment (SarahKirksey) [inline src/main/java/com/mraof/minestuck/entity/item/GristEntity.java:250]: And what?

File: src/main/java/com/mraof/minestuck/entity/item/GristEntity.java [hunk 8]
```diff
@@ -199,27 +243,67 @@ protected void readAdditionalSaveData(CompoundTag compound)
...
+				MSPacketHandler.sendToTracking(packet, this);
+				shaderAlpha = 255;
+			}
 		}
 	}
 	
+	
 	public void consumeGrist(PlayerIdentifier identifier, boolean sound)
 	{
+
```
Comment (SarahKirksey) [inline src/main/java/com/mraof/minestuck/entity/item/GristEntity.java:301]: These extra line breaks probably shouldn't be here.

File: src/main/java/com/mraof/minestuck/item/crafting/alchemy/GristSet.java [hunk 9]
```diff
@@ -27,8 +27,15 @@
 	
 	public static final GristSet EMPTY = new GristSet(Collections.emptyMap());
 
+	/**
+	this has been replaced to allow the GristGutter to have full access to this dictionary
+	 */
+	/*
 	private final Map<GristType, Long> gristTypes;
-
+	 */
```
Comment (SarahKirksey) [inline src/main/java/com/mraof/minestuck/item/crafting/alchemy/GristSet.java:35]: I think these comments can be safely removed

File: src/main/java/com/mraof/minestuck/client/gui/playerStats/EcheladderScreen.java (hunk 1 above)
Comment (Dweblenod) [reply src/main/java/com/mraof/minestuck/client/gui/playerStats/EcheladderScreen.java:197]: We use a formatting system for standardizing line spacing, brackets, etc. In IntelliJ you can go to the Code tab or you can press Ctrl + Alt + L while in a file to Reformat Code. Its definetely not a big deal but considering it takes a few seconds per file would you please do so

File: src/main/java/com/mraof/minestuck/client/gui/playerStats/EcheladderScreen.java (hunk 2 above)
Comment (Dweblenod) [reply src/main/java/com/mraof/minestuck/client/gui/playerStats/EcheladderScreen.java:242]: For the second part of your comment, see my comment on EcheladderScreen

File: src/main/java/com/mraof/minestuck/entity/item/GristEntity.java (hunk 8 above)
Comment (Dweblenod) [reply src/main/java/com/mraof/minestuck/entity/item/GristEntity.java:301]: See my comment on EcheladderScreen

**Comment by KingBalDoro** [comment]
hahahahahahahha el oh el i am the king of the github hahahah

File: src/main/java/com/mraof/minestuck/entity/item/GristEntity.java (hunk 7 above)
Comment (KingBalDoro) [reply src/main/java/com/mraof/minestuck/entity/item/GristEntity.java:250]: i forgor

File: src/main/java/com/mraof/minestuck/client/gui/playerStats/EcheladderScreen.java (hunk 3 above)
Comment (KingBalDoro) [reply src/main/java/com/mraof/minestuck/client/gui/playerStats/EcheladderScreen.java:282]: wuh

File: src/main/java/com/mraof/minestuck/client/renderer/entity/GristRenderer.java (hunk 4 above)
Comment (KingBalDoro) [reply src/main/java/com/mraof/minestuck/client/renderer/entity/GristRenderer.java:30]: yurp

Comment (KingBalDoro) [reply src/main/java/com/mraof/minestuck/client/renderer/entity/GristRenderer.java:30]: np[e

File: src/main/java/com/mraof/minestuck/entity/item/GristEntity.java (hunk 5 above)
Comment (KingBalDoro) [reply src/main/java/com/mraof/minestuck/entity/item/GristEntity.java:79]: honestly i think i'm gonna re comment all of our code since my commenting etiquite was off back when this was made

File: src/main/java/com/mraof/minestuck/entity/item/GristEntity.java (hunk 6 above)
Comment (KingBalDoro) [reply src/main/java/com/mraof/minestuck/entity/item/GristEntity.java:245]: understood

File: src/main/java/com/mraof/minestuck/entity/item/GristEntity.java (hunk 8 above)
Comment (KingBalDoro) [reply src/main/java/com/mraof/minestuck/entity/item/GristEntity.java:299]: done now die(jk i love you dweb you're my fave)

File: src/main/java/com/mraof/minestuck/item/crafting/alchemy/GristSet.java (hunk 9 above)
Comment (KingBalDoro) [reply src/main/java/com/mraof/minestuck/item/crafting/alchemy/GristSet.java]: done, how do i summon a pul request

**Comment by KingBalDoro** [comment]
nerd behavior but i like making the bot asay what i want it to say

**Comment by KingBalDoro** [comment]
check this out

**Comment by KingBalDoro** [comment]
i'm a dumb cat lololool i'm dumb

**Comment by KingBalDoro** [comment]
alright anyway, figuring out how this shit works, gonna take a while to get used to github. bear with me till i get it, gonna go through these edits the old fashion way till then. (you guys do such good work i love you)

File: src/main/java/com/mraof/minestuck/item/crafting/alchemy/GristTypes.java [hunk 10]
```diff
@@ -41,6 +41,7 @@
 	public static final RegistryObject<GristType> ARTIFACT = GRIST_TYPES.register("artifact", () -> new GristType(new GristType.Properties(0.1F, 1).candy(() -> MSItems.ARTIFACT_WARHEAD)));
 	public static final RegistryObject<GristType> ZILLIUM = GRIST_TYPES.register("zillium", () -> new GristType(new GristType.Properties(0.0F, 10).candy(() -> MSItems.ZILLIUM_SKITTLES).notUnderlingType()));
 	
+
```
Comment (Dweblenod) [inline src/main/java/com/mraof/minestuck/item/crafting/alchemy/GristTypes.java:44]: Try to avoid accidental line creation/removal like this. I typically check through everything I changed in the Commit changelist before going through with the commit to make sure I didnt do so. Its also good practice to make sure there arent any unused imports at the same time. In order to ensure I am adding changes that I wanted, and to document when I made the changes, I add a relevant Commit Message(also in the Commit window with the changelist)

File: src/main/java/com/mraof/minestuck/network/LotusFlowerPacket.java [hunk 11]
```diff
@@ -5,7 +5,8 @@
 import net.minecraft.network.FriendlyByteBuf;
 import net.minecraft.world.entity.Entity;
 
-public class LotusFlowerPacket implements PlayToClientPacket
+public class
+LotusFlowerPacket implements PlayToClientPacket
```
Comment (Dweblenod) [inline src/main/java/com/mraof/minestuck/network/LotusFlowerPacket.java:9]: Same thing as what I commented in GristTypes

File: src/main/java/com/mraof/minestuck/player/IdentifierHandler.java [hunk 12]
```diff
@@ -22,6 +22,8 @@
  */
 public class IdentifierHandler
 {
+	
+
```
Comment (Dweblenod) [inline src/main/java/com/mraof/minestuck/player/IdentifierHandler.java:26]: Same thing as what I commented in GristTypes

File: src/main/java/com/mraof/minestuck/world/storage/ClientPlayerData.java [hunk 13]
```diff
@@ -29,7 +29,7 @@
 	
 	private static Modus modus;
 	private static Title title;
-	private static int rung;
+	public static int rung;
```
Comment (Dweblenod) [inline src/main/java/com/mraof/minestuck/world/storage/ClientPlayerData.java:32]: was this change intentional?

File: src/main/java/com/mraof/minestuck/world/storage/PlayerSavedData.java [hunk 14]
```diff
@@ -99,7 +100,7 @@ public static PlayerData getData(PlayerIdentifier player, MinecraftServer server
 	{
 		return get(server).getData(player);
 	}
-
+
```
Comment (Dweblenod) [inline src/main/java/com/mraof/minestuck/world/storage/PlayerSavedData.java:103]: Same thing as what I commented in GristTypes

File: src/main/java/com/mraof/minestuck/entity/item/GristEntity.java [hunk 15]
```diff
@@ -59,6 +69,18 @@ public GristEntity(EntityType<? extends GristEntity> type, Level level)
...
+	 */
+	public GristEntity(Level level, double x, double y, double z, GristAmount gristData, int pickupDelay)
+	{
+		this(level, x, y, z, gristData);
+		consumeDelay = pickupDelay;
+		// Set the class's consume-delay variable to equal the pickupDelay value that got passed in.
+	}
+	
+	
+
```
Comment (Dweblenod) [inline src/main/java/com/mraof/minestuck/entity/item/GristEntity.java]: try to keep the line spacing consistent

**Review by Dweblenod — COMMENTED** [review]
Excited to see this Pull Request, and very proud of the work I see! This gets a rating of 5 yeehaws from me. I will check it out more thoroughly later, including in game. One major thing I will say which I think requires changing is the presence of the rung name code changes. Since it has nothing connecting it to the grist gutter that I have noticed so far, it is more appropriate to make a separate PR and remove the changes from this branch. That way it will be easier to review each. Please also make the title of the pull request more descriptive, its not immediately clear that this means a new system is being implemented

File: src/main/java/com/mraof/minestuck/entity/item/GristEntity.java (hunk 5 above)
Comment (Dweblenod) [reply src/main/java/com/mraof/minestuck/entity/item/GristEntity.java:79]: Go for it if you want, but this comment makes sense to me!

File: src/main/java/com/mraof/minestuck/world/storage/ClientPlayerData.java (hunk 13 above)
Comment (KingBalDoro) [reply src/main/java/com/mraof/minestuck/world/storage/ClientPlayerData.java:32]: idk

File: src/main/java/com/mraof/minestuck/client/gui/playerStats/EcheladderScreen.java (hunk 1 above)
Comment (KingBalDoro) [reply src/main/java/com/mraof/minestuck/client/gui/playerStats/EcheladderScreen.java:197]: done on all files. using this to test resolve feture

File: src/main/java/com/mraof/minestuck/entity/item/GristEntity.java (hunk 5 above)
Comment (KingBalDoro) [reply src/main/java/com/mraof/minestuck/entity/item/GristEntity.java:79]: oh yeah this was after i got it. those big blocks of text that just explain my thoughts on the feature are just annoying tho

File: src/main/java/com/mraof/minestuck/client/gui/playerStats/EcheladderScreen.java [hunk 16]
```diff
@@ -179,21 +186,62 @@ private void drawLadder(PoseStack poseStack, int currentRung, boolean showLastRu
...
+				{
+					if(I18n.exists("echeladder.rung." + rung))
+					{
+						s = I18n.get("echeladder.rung." + rung);
+					}
+					else
+					{
+						s = "Rung " + (rung + 1);
+					}
+				}
```
[the remark this answers is not in the document]
Comment (Dweblenod) [reply src/main/java/com/mraof/minestuck/client/gui/playerStats/EcheladderScreen.java:235]: This is basically what Sarah commented ^. If you make changes that resolve one comment related to these Strings it should resolve the other

File: src/main/generated/resources/assets/minestuck/lang/en_us.json [hunk 17]
```diff
@@ -944,6 +944,7 @@
   "echeladder.rung.7": "Imp Crusher",
   "echeladder.rung.8": "Slime Enthusiast",
   "echeladder.rung.9": "Burly Pocketeer",
+
```
[the remark this answers is not in the document]
Comment (KingBalDoro) [reply src/main/generated/resources/assets/minestuck/lang/en_us.json:947]: i'm not sure i understand exactly where to to do any of this but that may be the 4 glasses of champagne that i chugged

File: src/main/java/com/mraof/minestuck/client/gui/playerStats/EcheladderScreen.java (hunk 16 above)
[the remark this answers is not in the document]
Comment (KingBalDoro) [reply src/main/java/com/mraof/minestuck/client/gui/playerStats/EcheladderScreen.java:235]: interesting will check up when i'm less popped out of my head

File: src/main/java/com/mraof/minestuck/client/gui/playerStats/EcheladderScreen.java [hunk 18]
```diff
@@ -179,21 +186,62 @@ private void drawLadder(PoseStack poseStack, int currentRung, boolean showLastRu
...
-			String s = I18n.exists("echeladder.rung."+rung) ? I18n.get("echeladder.rung."+rung) : "Rung "+(rung+1);
-			mc.font.draw(poseStack, s, xOffset+ladderXOffset - mc.font.width(s) / 2, y + 2, textColor);
+			
+			String s = null;
+			if(playerTitle != null)
+			{
+				String playerBoots = String.valueOf((Minecraft.getInstance().player).getInventory().armor.get(0));
+				String playerLeggings = String.valueOf((Minecraft.getInstance().player).getInventory().armor.get(1));
+				String playerChestplate = String.valueOf((Minecraft.getInstance().player).getInventory().armor.get(2));
+				String playerHat = String.valueOf((Minecraft.getInstance().player).getInventory().armor.get(3));
```
[the remark this answers is not in the document]
Comment (KingBalDoro) [reply src/main/java/com/mraof/minestuck/client/gui/playerStats/EcheladderScreen.java:207]: didn't understand what you meant but now i do understand what you meant. honestly i think the aspect of "yeah it's customized bullshit" adds to it but i also see the issue in that it's never consistent. so what if we had a display of current session power level on the computer through some sburb program that shows you what exact level you and your fellow players are

File: src/main/java/com/mraof/minestuck/event/ServerEventHandler.java [hunk 19]
```diff
@@ -269,17 +266,60 @@ public static void onGetItemBurnTime(FurnaceFuelBurnTimeEvent event)
...
+
+	
 	@SubscribeEvent
 	public static void onPlayerTickEvent(TickEvent.PlayerTickEvent event)
 	{
 		if(!event.player.level.isClientSide)
 		{
 			PlayerData data = PlayerSavedData.getData((ServerPlayer) event.player);
+			
+			Player player = event.player;
```
[the remark this answers is not in the document]
Comment (KingBalDoro) [reply src/main/java/com/mraof/minestuck/event/ServerEventHandler.java:278]: ok

File: src/main/java/com/mraof/minestuck/item/crafting/alchemy/GristSet.java [hunk 20]
```diff
@@ -192,7 +199,38 @@ public GristSet addGrist(GristSet set)
...
+	public GristSet capGrist(int cap)
+	{
+		GristSet overflowGrist = new GristSet();
+		for(GristAmount amount : this.getAmounts())
+		{
+			if(amount.getAmount() > cap)
+			{
+				
+				System.out.println("grist is going to gutter");
```
[the remark this answers is not in the document]
Comment (KingBalDoro) [reply src/main/java/com/mraof/minestuck/item/crafting/alchemy/GristSet.java:213]: right ok makes sense

File: src/main/java/com/mraof/minestuck/network/LotusFlowerPacket.java (hunk 11 above)
Comment (KingBalDoro) [reply src/main/java/com/mraof/minestuck/network/LotusFlowerPacket.java:9]: u think this was resolved? i did the ctrl thingy multiple times

File: src/main/java/com/mraof/minestuck/world/storage/ClientPlayerData.java (hunk 13 above)
Comment (SarahKirksey) [reply src/main/java/com/mraof/minestuck/world/storage/ClientPlayerData.java:32]: It shouldn't have been, since you can do ClientPlayerData.getRung() instead of accessing it directly.

File: src/main/java/com/mraof/minestuck/item/crafting/alchemy/GristSet.java (hunk 20 above)
[the remark this answers is not in the document]
Comment (SarahKirksey) [reply src/main/java/com/mraof/minestuck/item/crafting/alchemy/GristSet.java:213]: In fact, it probably shouldn't be here at all.

**Comment by SarahKirksey** [comment]
> One major thing I will say which I think requires changing is the presence of the rung name code changes. Since it has nothing connecting it to the grist gutter that I have noticed so far, it is more appropriate to make a separate PR and remove the changes from this branch. That way it will be easier to review each.

While Dweblenod is right about this, it's important to note that separating features that are already together like this can actually be kinda tricky business. That's a major part of how things end up breaking in code cases, because you're separating things out and you accidentally forget something small but important. I'll try to help you with it, but it took me several years to figure out how to separate stuff out of my development branch, and I still frequently have to go back and make separate PRs to fix stuff that got broken during various splits and merges.

File: src/main/java/com/mraof/minestuck/entity/item/GristEntity.java (hunk 15 above)
Comment (KingBalDoro) [reply src/main/java/com/mraof/minestuck/entity/item/GristEntity.java]: my b

File: src/main/java/com/mraof/minestuck/item/crafting/alchemy/GristSet.java (hunk 20 above)
[the remark this answers is not in the document]
Comment (KingBalDoro) [reply src/main/java/com/mraof/minestuck/item/crafting/alchemy/GristSet.java:213]: i think it was a place holder for testing purposes i forgot to remove
